source/extensions/filters/common/rbac/engine_impl.cc

```diff
@@ -7,32 +7,47 @@ namespace Common {
 namespace RBAC {
 
 RoleBasedAccessControlEngineImpl::RoleBasedAccessControlEngineImpl(
-    const envoy::config::filter::http::rbac::v2::RBAC& config, bool disabled)
-    : engine_disabled_(disabled),
-      allowed_if_matched_(disabled ||
-                          config.rules().action() ==
-                              envoy::config::rbac::v2alpha::RBAC_Action::RBAC_Action_ALLOW) {
-  if (disabled) {
-    return;
-  }
+    const envoy::config::filter::http::rbac::v2::RBAC& config)
+    : config_(config) {
 
   for (const auto& policy : config.rules().policies()) {
     policies_.emplace_back(policy.second);
   }
+
+  for (const auto& policy : config.permissive_rules().policies()) {
+    permissive_policies_.emplace_back(policy.second);
+  }
 }
 
 RoleBasedAccessControlEngineImpl::RoleBasedAccessControlEngineImpl(
     const envoy::config::filter::http::rbac::v2::RBACPerRoute& per_route_config)
-    : RoleBasedAccessControlEngineImpl(per_route_config.rbac(), per_route_config.disabled()) {}
+    : RoleBasedAccessControlEngineImpl(per_route_config.rbac()) {}
 
 bool RoleBasedAccessControlEngineImpl::allowed(const Network::Connection& connection,
-                                               const Envoy::Http::HeaderMap& headers) const {
-  if (engine_disabled_) {
+                                               const Envoy::Http::HeaderMap& headers,
+                                               EnforcementMode mode) const {
+
+  // No enforced rule is set indicates RBAC isn't enabled for enforcement mode.
+  if ((mode == EnforcementMode::ENFORCED) && !config_.has_rules()) {
     return true;
   }
 
+  // No permissive rule is set indicates RBAC isn't enabled for permissive mode.
+  if ((mode == EnforcementMode::PERMISSIVE) && !config_.has_permissive_rules()) {
+    return true;
+  }
+
+  const std::vector<PolicyMatcher>& policies =
+      mode == EnforcementMode::PERMISSIVE ? permissive_policies_ : policies_;
+
+  bool allowed_if_matched = mode == EnforcementMode::PERMISSIVE
+                                ? config_.permissive_rules().action() ==
+                                      envoy::config::rbac::v2alpha::RBAC_Action::RBAC_Action_ALLOW
+                                : config_.rules().action() ==
+                                      envoy::config::rbac::v2alpha::RBAC_Action::RBAC_Action_ALLOW;
+
```

Comment:
This increases the complexity of the engine significantly. How about instead of encoding the knowledge of enforcement/permissiveness into the engine, we move that logically up to the filter, which instantiates two engines: one configured with the actual rules, and the other with the permissive ones? This would change the constructor for this to take the `config.rbac.v2alpha.RBAC` proto instead of the filter level one.

Reply:
+1

Reply:
done.